File: bindings/python/src/zylix_test/drivers/web.py
"""
Zylix Test Framework - Web Driver
"""

from typing import Any

from ..client import HttpClient
from ..types import WebDriverConfig
from .base import BaseDriver, BaseSession


class WebDriverSession(BaseSession[WebDriverConfig]):
    """Web driver session with browser-specific functionality."""

    def __init__(self, session_id: str, config: WebDriverConfig) -> None:
        """Initialize web session."""
        super().__init__(session_id, config)

    async def navigate_to(self, url: str) -> None:
        """Navigate to a URL.

        Args:
            url: URL to navigate to
        """
        await self._client.post(f"/session/{self._id}/url", {"url": url})

    async def get_url(self) -> str:
        """Get current URL.

        Returns:
            Current page URL
        """
        response = await self._client.get(f"/session/{self._id}/url")
        return str(response.get("value", ""))

    async def get_title(self) -> str:
        """Get page title.

        Returns:
            Current page title
        """
        response = await self._client.get(f"/session/{self._id}/title")
        return str(response.get("value", ""))

    async def execute_script(self, script: str, *args: object) -> Any:
        """Execute JavaScript in the browser.

        Args:
            script: JavaScript code to execute
            *args: Arguments to pass to the script

        Returns:
            Script execution result
        """
        response = await self._client.post(
            f"/session/{self._id}/execute/sync",
            {"script": script, "args": list(args)},
        )
        return response.get("value")

    async def back(self) -> None:
        """Navigate back in browser history."""
        await self._client.post(f"/session/{self._id}/back", {})

    async def forward(self) -> None:
        """Navigate forward in browser history."""
        await self._client.post(f"/session/{self._id}/forward", {})

    async def refresh(self) -> None:
        """Refresh the current page."""
        await self._client.post(f"/session/{self._id}/refresh", {})


class WebDriver(BaseDriver[WebDriverConfig, WebDriverSession]):
    """Web driver for browser automation."""

    def __init__(
        self,
        host: str = "127.0.0.1",
        port: int = 9515,
        timeout: int = 30000,
        browser: str = "chrome",
        headless: bool = False,
        viewport_width: int | None = None,
        viewport_height: int | None = None,
    ) -> None:
        """Initialize web driver.

        Args:
            host: Driver host address
            port: Driver port (default: 9515 for ChromeDriver)
            timeout: Request timeout in milliseconds
            browser: Browser type (chrome, firefox, safari, edge)
            headless: Run browser in headless mode
            viewport_width: Viewport width
            viewport_height: Viewport height
        """
        config = WebDriverConfig(
            host=host,
            port=port,
            timeout=timeout,
            browser=browser,  # type: ignore[arg-type]
            headless=headless,
            viewport_width=viewport_width,
            viewport_height=viewport_height,
        )
        super().__init__(config)

    async def create_session(self, **kwargs: Any) -> WebDriverSession:
        """Create a new browser session.

        Args:
            **kwargs: Additional session options

        Returns:
            New web session
        """
        # Merge config with kwargs
        browser = kwargs.get("browser", self._config.browser)
        headless = kwargs.get("headless", self._config.headless)

        capabilities: dict[str, Any] = {
            "capabilities": {
                "alwaysMatch": {
                    "browserName": browser,
                }
            }
        }

        # Add browser-specific options
        if browser == "chrome":
            chrome_options: dict[str, Any] = {"args": []}
            if headless:
                chrome_options["args"].append("--headless")
            capabilities["capabilities"]["alwaysMatch"]["goog:chromeOptions"] = chrome_options
        elif browser == "firefox":
            firefox_options: dict[str, Any] = {"args": []}
            if headless:
                firefox_options["args"].append("-headless")
            capabilities["capabilities"]["alwaysMatch"]["moz:firefoxOptions"] = firefox_options

        response = await self._client.post("/session", capabilities)
        value: dict[str, Any] = response.get("value", {})
        session_id = str(value.get("sessionId", ""))

        return WebDriverSession(session_id, self._config)

File: bindings/typescript/src/drivers/web.ts
/**
 * Zylix Test Framework - Web Driver
 */

import { BaseDriver, BaseSession } from './base.js';
import type { WebDriverConfig, WebSession } from '../types.js';

export class WebDriverSession extends BaseSession implements WebSession {
    constructor(id: string, config: WebDriverConfig) {
        super(id, config);
    }

    async navigateTo(url: string): Promise<void> {
        await this.client.post(`/session/${this.id}/url`, { url });
    }

    async getUrl(): Promise<string> {
        const response = await this.client.get(`/session/${this.id}/url`);
        return response.value as string;
    }

    async getTitle(): Promise<string> {
        const response = await this.client.get(`/session/${this.id}/title`);
        return response.value as string;
    }

    async executeScript<T>(script: string, args: unknown[] = []): Promise<T> {
        const response = await this.client.post(`/session/${this.id}/execute/sync`, {
            script,
            args,
        });
        return response.value as T;
    }

    async back(): Promise<void> {
        await this.client.post(`/session/${this.id}/back`, {});
    }

    async forward(): Promise<void> {
        await this.client.post(`/session/${this.id}/forward`, {});
    }

    async refresh(): Promise<void> {
        await this.client.post(`/session/${this.id}/refresh`, {});
    }
}

export class WebDriver extends BaseDriver<WebDriverConfig, WebDriverSession> {
    constructor(config: Partial<WebDriverConfig> = {}) {
        super({
            host: config.host ?? '127.0.0.1',
            port: config.port ?? 9515,
            timeout: config.timeout ?? 30000,
            browser: config.browser ?? 'chrome',
            headless: config.headless ?? false,
            viewportWidth: config.viewportWidth ?? 1920,
            viewportHeight: config.viewportHeight ?? 1080,
        });
    }

    async createSession(options?: Partial<WebDriverConfig>): Promise<WebDriverSession> {
        const mergedConfig = { ...this.config, ...options };

        const capabilities: Record<string, unknown> = {
            capabilities: {
                alwaysMatch: {
                    browserName: mergedConfig.browser,
                },
            },
        };

        // Add Chrome-specific options
        if (mergedConfig.browser === 'chrome') {
            const chromeOptions: Record<string, unknown> = {
                args: [],
            };

            if (mergedConfig.headless) {
                (chromeOptions.args as string[]).push('--headless=new');
            }

            if (mergedConfig.viewportWidth && mergedConfig.viewportHeight) {
                (chromeOptions.args as string[]).push(
                    `--window-size=${mergedConfig.viewportWidth},${mergedConfig.viewportHeight}`
                );
            }

            (capabilities.capabilities as Record<string, unknown>).alwaysMatch = {
                ...(capabilities.capabilities as Record<string, unknown>).alwaysMatch as object,
                'goog:chromeOptions': chromeOptions,
            };
        }

        const response = await this.client.post('/session', capabilities);
        const value = response.value as Record<string, string>;
        const sessionId = value.sessionId;

        return new WebDriverSession(sessionId, mergedConfig);
    }
}
